homeassistant/components/wled/sensor.py

```diff
@@ -0,0 +1,132 @@
+    DOMAIN,
+    TIME_SECONDS,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType,
+    entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity], bool], None],
+) -> None:
+    """Set up WLED sensor based on a config entry."""
+    wled: WLED = hass.data[DOMAIN][entry.entry_id][DATA_WLED_CLIENT]
+
+    sensors = [
+        WLEDEstimatedCurrentSensor(entry.entry_id, wled),
+        WLEDUptimeSensor(entry.entry_id, wled),
+        WLEDFreeHeapSensor(entry.entry_id, wled),
+    ]
+
+    async_add_entities(sensors, True)
+
+
+class WLEDSensor(WLEDDeviceEntity):
+    """Defines a WLED sensor."""
+
+    def __init__(
+        self,
+        entry_id: str,
+        wled: WLED,
+        name: str,
+        icon: str,
+        unit_of_measurement: str,
+        key: str,
+    ) -> None:
+        """Initialize WLED sensor."""
+        self._state = None
+        self._unit_of_measurement = unit_of_measurement
+        self._key = key
+
+        super().__init__(entry_id, wled, name, icon)
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return f"{self.wled.device.info.mac_address}_{self._key}"
+
+    @property
+    def state(self) -> Union[None, str, int, float]:
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit this state is expressed in."""
+        return self._unit_of_measurement
+
+
+class WLEDEstimatedCurrentSensor(WLEDSensor):
+    """Defines a WLED estimated current sensor."""
+
+    def __init__(self, entry_id: str, wled: WLED) -> None:
+        """Initialize WLED estimated current sensor."""
+        super().__init__(
+            entry_id,
+            wled,
+            f"{wled.device.info.name} Estimated Current",
+            "mdi:power",
+            CURRENT_MA,
+            "estimated_current",
+        )
+
+    async def _wled_update(self) -> None:
+        """Update WLED entity."""
+        self._state = self.wled.device.info.leds.power
+        self._attributes = {
+            ATTR_LED_COUNT: self.wled.device.info.leds.count,
+            ATTR_MAX_POWER: self.wled.device.info.leds.max_power,
+        }
+
+
+class WLEDUptimeSensor(WLEDSensor):
+    """Defines a WLED uptime sensor."""
+
+    def __init__(self, entry_id: str, wled: WLED) -> None:
+        """Initialize WLED uptime sensor."""
+        super().__init__(
+            entry_id,
+            wled,
+            f"{wled.device.info.name} Uptime",
+            "mdi:clock-outline",
+            TIME_SECONDS,
+            "uptime",
+        )
+
+    async def _wled_update(self) -> None:
+        """Update WLED uptime sensor."""
+        self._state = self.wled.device.info.uptime
```

Comment:
We don't allow relative times in the state machine since that can spam updates just due to time passing.

Please implement the timestamp sensor device class and calculate and format the time to absolute utc isoformatted timestamp. This sensor should be unit less.

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

The user will be able to show this timestamp as relative time in the lovelace card.

Reply:
I’ve also read that we should not change the data provided by the device. So which one is it now?

Reply:
The relative time rule trumps the other rule.

Reply:
I see the importance for the state engine will adjust 👍